Change save animation on cookies page

MatthewWilkes opened the pull request:
The current save animation causes the save button to expand with additional text, then ease back to the original. This replaces the button contents with a saving message that causes less visual disturbance.

The current version appears as:
![croydon](https://user-images.githubusercontent.com/98908/222109806-54b7762b-8774-4eb6-8b39-35250003d5a1.gif)

with this change it looks as follows:
![essex](https://user-images.githubusercontent.com/98908/222109872-865824d6-80fc-4a61-a843-164419d22add.gif)

*N.B.:* Differences in accent colour and other styling are due to this being captured on different sites, only the button behaviour is impacted.

AnthonySGeorge committed:
Change save animation on cookies page

The current save animation causes the save button to expand with additional text, then ease back to the original. This replaces the button contents with a saving message that causes less visual disturbance.

MatthewWilkes commented:
FYI @Adnan-cds

Adnan-cds commented:
Thanks for the code change (and particularly for the GIFs) Matthew.  Looks okay to me.  Chris, Finn, Mark, have you got any opinion? Copying @cjstevens78  @finnlewis @markconroy

andybroomfield reviewed (approved):
I'm happy with this change as it will improve the usuability of the button which was a common issue raised.

andybroomfield commented:
Have approved this PR.
I also think we should look at how we can make this configurable. Ideally BHCC would like to be able to get it to submit as a page reload and disaplay a message to be consitent with our current forms.
BHCC currently patches the module to to something like this (still JS based).

<img width="954" alt="Screenshot 2023-03-01 at 12 38 11 pm" src="https://user-images.githubusercontent.com/1467480/222142249-0e824f4b-be5f-4408-84b1-700e7dddc07c.png">

keelanfh commented:
I like the BHCC pattern @andybroomfield, it would be interesting to see in time if we could move to something like that as a default for this module. But the spinner is OK for now.

markconroy commented on js/localgov_eu_cookie_compliance.es6.js:
```diff
@@ -103,10 +103,24 @@
+      setTimeout(
+        function () {
+          saveButton.classList.remove('spinning');
+          saveButton.innerHTML = `${Drupal.t("Preferences saved")}  <i class="fas fa-check"></i>`;
```
I'm nervous of these classes. Not all themes will have font-awesome installed. I'd guess most themes will not have it if they are sub themes of Localgov Base which I think most councils are using for their base theme.

We should rewrite these using BEM-style classes. I also have [a larger issue](https://github.com/localgovdrupal/localgov_eu_cookie_compliance/issues/2) to do that for this whole module.

MatthewWilkes replied:
I'd initially filtered these out of the PR compared to what we were doing internally, but then noticed that the module as a whole was using them. I think consistency is really important here, so if #2 is broadly acceptable then we can aim this for the same release that makes that change. I'd feel a bit strange about removing these while the core templates include the checks.

markconroy replied:
Let's leave them in for now, and we'll fix all those classes on one go via #2 when we get around to it.

It might mean leaving in the bootstrap classes so anyone currently using this module won't have a breaking change and additionally adding our new BEM classes.

markconroy commented on css/save-animation.css:
```diff
@@ -0,0 +1,34 @@
+    padding-right: 40px;
+}
+
+.eu-cookie-compliance-save-preferences-button.spinning:before {
```
Can we use double colons for pseudo elements - `::before` to follow Drupal's coding standards.

markconroy commented on js/localgov_eu_cookie_compliance.es6.js:
```diff
@@ -103,10 +103,24 @@
-        .appendTo(e.target)
-        .hide(2000);
-    });
+    var saveButton = document.querySelector('.eu-cookie-compliance-save-preferences-button');
```
Let's change this `var` to a `const`

markconroy commented on js/localgov_eu_cookie_compliance.es6.js:
```diff
@@ -103,10 +103,24 @@
+    }, false);
+
   }
 })(jQuery, Drupal);
```
Is jQuery needed here? It would be good to not have a jQuery dependency in this module if possible.

MatthewWilkes replied:
Agreed, but as this isn't a change in this PR I think we should open a new issue to drop the jQuery dep.

markconroy replied:
Good idea. Can you create that issue please?

Adnan-cds replied:
I have raised a [ticket](https://github.com/localgovdrupal/localgov_eu_cookie_compliance/issues/6) to this effect.

markconroy commented on css/save-animation.css:
```diff
@@ -0,0 +1,34 @@
+.eu-cookie-compliance-save-preferences-button {
```
Let's give the css file name something that resembles the module - `eu-cookie-compliance.css` or something similar. I can imagine we will have CSS in here that is not just about the animation.

finnlewis commented:
Awaiting minor changes from @MatthewWilkes if poss :)

MatthewWilkes commented:
Will do, when back from holiday!

finnlewis commented:
Leaving this until @MatthewWilkes gets back from holiday! (stop checking your GitHub notifications @MatthewWilkes !)

MatthewWilkes committed:
Address PR comments

This changes the name of the CSS file to something more generic and fixes some style issues in JS and CSS.

MatthewWilkes committed:
Fix colour of spinner to match text

The colour was previously set to be the accent colour, but this is because our site has --button-text-color-hover set to the accent colour. It works less well on Croydon (and others). This change makes the spinner bulk match the text.

There is a problem here, as the rest of the circle is a 10% opacity version of this colour, but we can't achieve that with CSS variables until CSS values and units 5 is supported (to allow for extracting the r,g,b tuple from the hex and reinserting into an rgba). As such, our CSS currently sets this a 10% red, but this is an artifact of Essex. We can achieve a similar (but aliased) effect by setting it to the colour but 0.1px width, but to my eyes this looks more wrong than the circle being the wrong colour given how faint it is. Personally, I don't think the red is sufficiently visable to be recognisable, but I'd take opinions on this.

MatthewWilkes commented on css/eu-cookie-compliance.css:
```diff
@@ -0,0 +1,34 @@
+    border-radius: 50%;
+    right: 6px;
+    top: 50%;
+    box-shadow: 0 0 0 1px rgba(255, 0, 0, 0.1);
```
There is a problem here, as the rest of the circle is a 10% opacity version of the main colour, but we can't achieve that with CSS variables until CSS values and units 5 is supported (to allow for extracting the r,g,b tuple from the hex and reinserting into an rgba). As such, our CSS currently sets this a 10% red, but this is an artifact of Essex. We can achieve a similar (but aliased) effect by setting it to the colour but 0.1px width, but to my eyes this looks more wrong than the circle being the wrong colour given how faint it is. Personally, I don't think the red is sufficiently visable to be recognisable, but I'd take opinions on this.

Desiredlook (with Croydon colours):
![image](https://user-images.githubusercontent.com/98908/227602006-7b3fc107-d431-4b5e-bb8c-1dfee491fc31.png)

Current:
![image](https://user-images.githubusercontent.com/98908/227602138-e9032d26-0540-481f-b659-6e8d67c9b992.png)

0.1px version:
![image](https://user-images.githubusercontent.com/98908/227602281-0a46665d-ea59-4091-b04a-6e10b4840a88.png)

Any ideas @markconroy? Adding @AnthonySGeorge as this was his change, he may have opinions.

MatthewWilkes replied:
Actually, this might work?

`box-shadow: 0 0 2px 0 var(--button-text-color-hover);`
![image](https://user-images.githubusercontent.com/98908/227603380-0083098e-7680-4a30-ab73-deedeaf2bfbb.png)

MatthewWilkes commented on css/eu-cookie-compliance.css:
```diff
@@ -0,0 +1,34 @@
+    right: 6px;
+    top: 50%;
+    box-shadow: 0 0 0 1px rgba(255, 0, 0, 0.1);
+    border-right: 3px solid var(--button-text-color-hover);
```
@markconroy FYI, I've changed this variable to one that seems more appropriate.

MatthewWilkes commented:
@finnlewis Resolved the three comments, however I've tagged @markconroy in a question I noticed while fixing them.

MatthewWilkes committed:
Change box shadow of the spinner to avoid transparency

This changes the box shadow to avoid needing an alpha channel in the colour, as that can't be extracted from the hex code.

markconroy commented:
2 items

1. I'm not sure what the issue we are fixing with the colour/circle is
2. Using CSS variables in modules (especially if they are referring to variables that are only found in localgov_base theme) has the potential to be problematic if any one uses this module but is not also using the localgov_base theme. It might be better to hardcode a value here and then we can replicate it in the localgov_base theme to make it easier to override.

MatthewWilkes commented:
@markconroy So, the issue with the colour is that it should be the same colour as the text, so it's visible against the background. It was previously using a combination of the accent colour through a variable and 10% opacity red, this latest change makes it use the button text colour.

Happy to hard-code a value here for overriding elsewhere, do you have a preference? For Essex we want pure red, for Croydon white would make most sense.

markconroy commented:
Thanks @MatthewWilkes 

I guess white is probably best for now, so it contrasts against the background as much as possible.

MatthewWilkes commented:
@markconroy Can do, but there's no guarantee that it'll have much contrast against the background of the button, as it might itself be white or light-coloured. I'll also add a PR against the base theme to add the parts of this CSS rule that include the variables. I might need some guidance as to what the selector should be, as `.eu-cookie-compliance-save-preferences-button.spinning::before` is already pretty specific.

finnlewis commented:
Leaving this for discussion and conclusion.

markconroy commented:
Hi @MatthewWilkes 

Just checking if you have this updated, or if we can get it updated so we can have this issue merged and closed.

MatthewWilkes commented:
@markconroy If I'm honest, I'm not 100% on what we decided to do after that chat in slack. Am I setting this white and putting a reference to it in the base theme?

markconroy commented:
I'll be honest, I can't remember either 🤣 

I think the basic thing is "do whatever needs hard-coding colours for now in the module, so we can get it released" and we'll have a more scalable approach in the localgov_base theme (using variables).

MatthewWilkes committed:
Remove use of theme variables

This removes the use of LGD theme base variables, setting the values to white instead. It includes the original rules in a comment, for ease of including in LGD base.

MatthewWilkes commented:
Pushed that change

markconroy reviewed (changes requested):
Let's remove the commented out code. It's not good practice to have it there, and in the current state it looks like we are encouraging people to uncomment these lines if using localgov_base (or at least that's how I read it) - but if they do that, their changes will be overridden when they do an update.

MatthewWilkes committed:
Remove LGD base theme info

This removes the comment about how to integrate with the LGD base theme.

MatthewWilkes commented:
Removed that comment

finnlewis reviewed (approved):
Looks good to me and changes have been made since @markconroy's comments.
